syntax = "proto3";

service Chat {
    rpc connect(ChatUser) returns (ChatUserConnected);
    rpc disconnect(ChatUserConnected) returns (ChatUserConnected);
    rpc sendMessage(ChatMessage) returns (ServerResponse);
    rpc subscribeMessages(ChatUserConnected) returns (stream ChatMessage);
    rpc subscribeActiveUsers(ChatUserConnected) returns (stream ChatActiveUser);
    rpc register(Creds) returns (ServerResponse);
    rpc login(Creds) returns (ServerResponse);
    rpc getAllUsers(Empty) returns (ServerResponse);
    rpc createGroup(CreateGroupRequest) returns (ServerResponse);
    rpc displayMemberGroups(ChatUser) returns (ServerResponse);
}

message CreateGroupRequest{
    string username = 1;
    string group_name = 2;
    string group_members = 3;
}

message Empty{}

message ChatUser {
    string username = 1;
}

message ChatUserConnected {
    //string userId = 1;
    string username = 1;
}

message ChatActiveUser {
    string username = 1;
    string currentHash = 2;
}

message ChatUserDisconnect {
    bool isDisconnected = 1;
}

message ChatMessage {
    string username = 1;
    string message = 2;
    string recipient = 3;
    bool is_broadcast = 4;
    bool is_group_chat_msg = 5;
    string group_name = 6;

}

message Creds {
    string username = 1;
    string password = 2;
}

message ServerResponse {
    string response = 1;
}